### PR TITLE
 [PropertyInfo] Check if the virtual property has a setter in `isAllowedProperty`

### LINKED ISSUE #58556: [PropertyInfo] PropertyInfo is not compatible with PHP 8.4 property hooks
### Symfony version(s) affected

7.1.4

### Description

In PHP 8.4 we'll have virtual properties and asymmetric visibility.

PropertyInfo considers the following properties writable, although they are not:

```php
final class Foo
{
    public private(set) bool $public_private_set;
    public protected(set) bool $public_protected_set;
    public bool $virtual_no_set_hook { get => true; }
}
```

A side note. Many Symfony components optimistically state that they are compatible with all future PHP versions (for instance, the PropertyInfo component has `"require": { "php": ">=8.2" }` in its `composer.json`). However, this case shows that they are not and cannot be. I've discussed, whether it is a BC break on PHP's side or it isn't, see https://externals.io/message/125772. Symfony should probably reconsider it's PHP constraints.

### How to reproduce

https://github.com/vudaltsov/symfony-property-access-php84

### Possible Solution

In PHP 8.4 a property is writable if 
```php
   $reflectionProperty->isPublic()
&& !$reflectionProperty->isReadonly()
&& !$reflectionProperty->isPrivateSet()
&& !$reflectionProperty->isProtectedSet()
&& (!$reflectionProperty->isVirtual() || $reflectionProperty->hasHook(PropertyHookType::Set))
```

See https://externals.io/message/125740.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by pan93412**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | 
| Issues        | Fix #58556
| License       | MIT

Separate from #58960

### CODE REVIEW

**Comment by pan93412** [comment]
cc @xabbuh

File: src/Symfony/Component/PropertyInfo/Tests/Fixtures/VirtualProperties.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  * file that was distributed with this source code.
+  */
+
+namespace Symfony\Component\PropertyInfo\Tests\Fixtures;
+
+class VirtualProperties
+{
+    public bool $virtualNoSetHook { get => true; }
+    public bool $virtualSetHookOnly { set => $value; }
+    public bool $virtualHook { get => true; set => $value; }
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Tests/Fixtures/VirtualProperties.php]: maybe we should add tests with non-virtual property hooks?

Comment (pan93412) [reply src/Symfony/Component/PropertyInfo/Tests/Fixtures/VirtualProperties.php]: Yeah, I forgot there is a backed property hooks when I wrote down this code. Let me fix it.

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php [hunk 2]
```diff
@@ -625,6 +625,14 @@ private function isAllowedProperty(string $class, string $property, bool $writeA
                 return false;
             }
 
+            if (\PHP_VERSION_ID >= 80400
+                    && $writeAccessRequired
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: I think we can nest all "if"s here in `if ($writeAccessRequired) {`

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php [hunk 3]
```diff
@@ -625,6 +625,14 @@ private function isAllowedProperty(string $class, string $property, bool $writeA
...
+                    && $reflectionProperty->isVirtual()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: now I'm wondering: why excluding virtual properties?

Comment (pan93412) [reply src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: A backed property can be set even there is no set hook.
A virtual property cannot be set if there is no set hook.

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php (hunk 2 above)
Comment (pan93412) [reply src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: Good idea. Nested: https://github.com/symfony/symfony/pull/58962/commits/4b89fc082fad05385e019b5c5e1931975dd5dbef

**Comment by pan93412** [comment]
Superseded by #58963.

**Pull request closed without merging**